src/main/java/graphql/execution/instrumentation/fieldvalidation/FieldValidationSupport.java

```diff
@@ -0,0 +1,185 @@
+package graphql.execution.instrumentation.fieldvalidation;
+
+import graphql.ErrorType;
+import graphql.GraphQLError;
+import graphql.Internal;
+import graphql.analysis.QueryTraversal;
+import graphql.analysis.QueryVisitorEnvironment;
+import graphql.execution.ExecutionContext;
+import graphql.execution.ExecutionPath;
+import graphql.language.Field;
+import graphql.language.SourceLocation;
+import graphql.schema.GraphQLCompositeType;
+import graphql.schema.GraphQLFieldDefinition;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+@Internal
+class FieldValidationSupport {
+
+    static List<GraphQLError> validateFieldsAndArguments(FieldValidation fieldValidation, ExecutionContext executionContext) {
+
+        Map<ExecutionPath, FieldAndArguments> fieldArgumentsMap = new HashMap<>();
+
+        QueryTraversal queryTraversal = new QueryTraversal(
+                executionContext.getGraphQLSchema(),
+                executionContext.getDocument(),
+                executionContext.getOperationDefinition().getName(),
+                executionContext.getVariables()
+        );
+        queryTraversal.visitPreOrder(traversalEnv -> {
+            Field field = traversalEnv.getField();
+            if (field.getArguments() != null && !field.getArguments().isEmpty()) {
+                //
+                // only fields that have arguments make any sense to placed in play
+                // since only they have variable input
+                FieldAndArguments fieldArguments = new FieldAndArgumentsImpl(traversalEnv);
+                fieldArgumentsMap.put(fieldArguments.getPath(), fieldArguments);
+            }
+        });
+
+        FieldValidationEnvironment environment = new FieldValidationEnvironmentImpl(executionContext, fieldArgumentsMap);
+        //
+        // this will allow a consumer to plugin their own validation of fields and arguments
+        return fieldValidation.validateFields(environment);
+    }
+
+    private static class FieldAndArgumentsImpl implements FieldAndArguments {
+        private final QueryVisitorEnvironment traversalEnv;
+        private final FieldAndArguments parentArgs;
+        private final ExecutionPath path;
+
+        FieldAndArgumentsImpl(QueryVisitorEnvironment traversalEnv) {
+            this.traversalEnv = traversalEnv;
+            this.parentArgs = mkParentArgs(traversalEnv);
+            this.path = mkPath(traversalEnv);
+        }
+
+        private FieldAndArguments mkParentArgs(QueryVisitorEnvironment traversalEnv) {
+            return traversalEnv.getParentEnvironment() != null ? new FieldAndArgumentsImpl(traversalEnv.getParentEnvironment()) : null;
+        }
+
+        private ExecutionPath mkPath(QueryVisitorEnvironment traversalEnv) {
```

Comment:
not sure ExecutionPath is unique for queries including fragments:
for example:  
``` graphql
{foo{ 
... on Bar1 {name}
... on Bar2 {name}
}}

QueryTraversal visits every possible outcome: This means name is visited for `Bar1` and for `Bar2`.

Reply:
Ahh you are right. A quick test reveals

```
        def query = '''
        query VarTest($stringVar : String, $intVar : Int, $inputDataVar : InputData, $idVar : ID, $fmt1Var : String) {
            field1(inputData : $inputDataVar, id : $idVar) {
                ...fragOnField 
                ... on Information {  informationString (fmt1 : "inlineValue") }
            }
        }
        
        fragment fragOnField on Information {
            informationString(fmt1 : $fmt1Var)
        }
        '''
```

has 2 fields called logically `\field1\informationString` both with different values.  Hmmm

One way would be for the map to be Map<ExecutionPath><List<FieldAndArgs>> where we have mostly 1 item but there can be N.

Another though is that is becomes just a <List<FieldAndArgs>> and you have to iterate thought them yourself.  They all have `path` attached inside them so they can be address more meaningfully than just "field name"

What other thoughts do you have on how we can address this design challenge.

Reply:
Thinking about it more - why not do both?

Thats what I did

Reply:
side note: I am not sure about your example: it should fail validation because `informationString` must be exactly the same field (including arguments) because at runtime it is only one field.

I think the only time ExecutionPath is not unique is when you have two ore more fragments with different type conditions and both have have subfields with the same name/alias.

But how to solve the problem: I think a Map ExecutionPath -> List is a good solution. But sure: we can of course do both.

Reply:
Yup - full  validation gives

`Validation error of type FieldsConflict: alias2: they have differing arguments`